pkg/image/origins/main.go

```diff
@@ -0,0 +1,188 @@
+	for k := range img.OriginMap {
+		_, ok := currentImages[k]
+		if !ok {
+			unusedImages = append(unusedImages, k)
+		}
+	}
+
+	return unusedImages
+}
+
+func PrintUpdatedImageOrigins(linuxImagesFromArgs, targetImages, targetWindowsImages []string) error {
+	fmt.Println("Generating updated rancher-image-origins map")
+
+	// use the existing map so that we don't
+	// check images which have already been resolved.
+	imgToURL := img.OriginMap
+
+	// look through any images passed as arguments
+	err := convertImagesToRepoUrls(linuxImagesFromArgs, imgToURL)
+	if err != nil {
+		return err
+	}
+	// look through the linux target images
+	err = convertImagesToRepoUrls(targetImages, imgToURL)
+	if err != nil {
+		return err
+	}
+	// look through the windows target images
+	err = convertImagesToRepoUrls(targetWindowsImages, imgToURL)
+	if err != nil {
+		return err
+	}
+
+	// order images alphabetically
+	type sorter struct {
+		img string
+		url string
+	}
+	var sorted []sorter
+	for k, v := range imgToURL {
+		sorted = append(sorted, sorter{img: k, url: v})
+	}
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].img < sorted[j].img
+	})
+
+	x := "OriginMap = map[string]string{\n"
+	for _, e := range sorted {
+		x = x + "	\"" + e.img + "\":" + " " + "\"" + e.url + "\"," + "\n"
+	}
+	x = x + "}\n"
+
+	// print result, can be copied into pkg/image/origins.go
+	// to automatically update some images.
+	fmt.Println(x)
+
+	// find all images that have not been resolved yet
+	var unknownImages string
+	for k, v := range imgToURL {
+		if v == "" || v == "unknown" {
+			unknownImages = unknownImages + "\n" + k
+		}
+	}
+
+	// warn about unresolved images
+	// so that they may be manually resolved
+	if len(unknownImages) > 0 {
+		fmt.Println(fmt.Sprintf("[WARN] Some images do not have an origin defined, please provide origins within rancher/pkg/image/origins.go for the following images: %s", unknownImages))
+	}
+
+	return nil
+}
+
+// convertImagesToRepoUrls attempts to resolve Rancher owned images with its origin repository.
+// It does so on a best-effort basis, any gaps left by this must be covered manually
+// in the OriginMap in origins.go
+func convertImagesToRepoUrls(images []string, imgToURL map[string]string) error {
+	for _, repo := range img.UniqueTargetImages(images) {
+		if v, ok := imgToURL[repo]; ok && (v != "" && v != "unknown") {
+			continue // image is already resolved
+		}
+		switch {
+		case strings.Contains(repo, "mirrored"):
+			// no way to automatically determine origin.
+			// needs manual resolution, see origins.go.
+			imgToURL[repo] = "unknown"
+		case strings.Contains(repo, "hardened"):
+			// hardened images conversion, best effort
+			cleanRepo := strings.ReplaceAll(repo, "hardened-", "image-build-")
+			url, err := checkURL("https://www.github.com/rancher/" + cleanRepo)
+			if err != nil && err.Error() != imageNotFound {
+				return err
+			}
+			if err != nil {
+				url = "unknown"
+			}
+			imgToURL[repo] = url
```

Comment:
```suggestion
		case strings.Contains(repo, "hardened"):
			// hardened images conversion, best effort
			repo = strings.ReplaceAll(repo, "hardened-", "image-build-")
			fallthrough
```
I'm not saying that you have to, just that it's possible.

Reply:
Yeah, I had thought about that too its just that some of the hardened images have slightly different names than their base repo's, for instance the `hardened-dns-node-cache` comes from `https://github.com/rancher/image-build-dns-nodecache`, which doesn't have a `-` separating `node` and `cache`.

I'll do a double check of the hardened images tomorrow to make sure I have them all correct, the list still isn't completely finalized